File: test/interop-clab/vendor_images.py
#!/usr/bin/env -S uv run --frozen --no-default-groups -- python
"""Prepare vendor-specific images for the Go-owned containerlab bootstrap."""

from __future__ import annotations

import argparse
import json
import logging
import os
import shutil

# All subprocess calls use fixed argv and an allowlisted executable resolved to
# an absolute path. Shell execution is never enabled.
import subprocess  # nosec B404
import tarfile
import tempfile
from pathlib import Path
from typing import TYPE_CHECKING, Any, override
from urllib.error import URLError
from urllib.parse import urlsplit
from urllib.request import HTTPRedirectHandler, Request, build_opener

if TYPE_CHECKING:
    from http.client import HTTPMessage
    from typing import IO

VYOS_GITHUB_API = "https://api.github.com/repos/vyos/vyos-rolling-nightly-builds/releases"
VYOS_DOWNLOAD_BASE = "https://github.com/vyos/vyos-rolling-nightly-builds/releases/download"
_ALLOWED_HTTPS_HOSTS = frozenset({"api.github.com", "github.com"})
_ALLOWED_EXECUTABLES = frozenset(
    {
        "7z",
        "podman",
        "tar",
        "unsquashfs",
        "xorriso",
    },
)
_log = logging.getLogger("vendor-images")


class _VendorSecurityError(ValueError):
    """Report a rejected URL or executable boundary."""


class _ExecutableNotFoundError(FileNotFoundError):
    """Report an unavailable executable after allowlist validation."""


def _validated_https_url(url: str) -> str:
    """Return *url* only when it uses the exact outbound HTTPS allowlist."""
    parsed = urlsplit(url)
    try:
        port = parsed.port
    except ValueError as exc:
        message = f"invalid outbound URL port: {url!r}"
        raise _VendorSecurityError(message) from exc
    if (
        parsed.scheme != "https"
        or parsed.hostname not in _ALLOWED_HTTPS_HOSTS
        or parsed.username is not None
        or parsed.password is not None
        or port not in (None, 443)
    ):
        message = f"outbound URL is not allowlisted: {url!r}"
        raise _VendorSecurityError(message)
    return url


class _AllowlistedRedirectHandler(HTTPRedirectHandler):
    """Reject redirects outside the exact outbound HTTPS allowlist."""

    @override
    def redirect_request(
        self,
        req: Request,
        fp: IO[bytes],
        code: int,
        msg: str,
        headers: HTTPMessage,
        newurl: str,
    ) -> Request | None:
        """Validate a redirect target before constructing its request."""
        _validated_https_url(newurl)
        return super().redirect_request(req, fp, code, msg, headers, newurl)


def _resolve_executable(command: str) -> str:
    """Resolve an allowlisted command to an absolute executable path."""
    if command not in _ALLOWED_EXECUTABLES:
        message = f"executable is not allowlisted: {command!r}"
        raise _VendorSecurityError(message)
    found = shutil.which(command)
    if found is None:
        message = f"allowlisted executable not found: {command}"
        raise _ExecutableNotFoundError(message)
    resolved = Path(found)
    if not resolved.is_file() or not os.access(resolved, os.X_OK):
        message = f"allowlisted executable is not executable: {resolved}"
        raise _ExecutableNotFoundError(message)
    return str(resolved.absolute())


def _run(
    command: list[str],
    *,
    check: bool = True,
    capture: bool = False,
    dry_run: bool = False,
) -> subprocess.CompletedProcess[str]:
    """Execute one fixed-argv allowlisted command."""
    if not command:
        message = "command must not be empty"
        raise _VendorSecurityError(message)
    if dry_run:
        if command[0] not in _ALLOWED_EXECUTABLES:
            message = f"executable is not allowlisted: {command[0]!r}"
            raise _VendorSecurityError(message)
        _log.info("[dry-run] %s", " ".join(command))
        return subprocess.CompletedProcess(command, 0, stdout="", stderr="")
    resolved = [_resolve_executable(command[0]), *command[1:]]
    return subprocess.run(  # nosec B603  # noqa: S603
        resolved,
        check=check,
        text=True,
        capture_output=capture,
    )


def _image_exists(reference: str, *, dry_run: bool = False) -> bool:
    """Return whether an image reference exists in local Podman storage."""
    if dry_run:
        return False
    result = _run(
        ["podman", "image", "exists", reference],
        check=False,
        capture=True,
    )
    return result.returncode == 0


def _which(name: str) -> str | None:
    """Locate an executable on PATH."""
    return shutil.which(name)


def _latest_vyos_version() -> str:
    """Fetch the latest VyOS rolling release tag from GitHub API."""
    request = Request(  # noqa: S310
        _validated_https_url(f"{VYOS_GITHUB_API}/latest"),
        headers={
            "Accept": "application/vnd.github+json",
            "User-Agent": "gobfd-bootstrap",
        },
    )
    try:
        opener = build_opener(_AllowlistedRedirectHandler())
        with opener.open(request, timeout=30) as response:
            data: dict[str, Any] = json.loads(response.read())
    except (URLError, json.JSONDecodeError) as exc:
        message = f"query VyOS releases: {exc}"
        raise RuntimeError(message) from exc
    tag = data.get("tag_name", "")
    if not isinstance(tag, str) or not tag:
        message = "query VyOS releases: response lacks tag_name"
        raise RuntimeError(message)
    return tag


def _download_vyos_iso(version: str, destination: Path) -> Path:
    """Download one VyOS rolling ISO."""
    filename = f"vyos-{version}-generic-amd64.iso"
    url = f"{VYOS_DOWNLOAD_BASE}/{version}/{filename}"
    target = destination / filename
    request = Request(  # noqa: S310
        _validated_https_url(url),
        headers={"User-Agent": "gobfd-bootstrap"},
    )
    try:
        opener = build_opener(_AllowlistedRedirectHandler())
        with opener.open(request, timeout=600) as response, target.open("wb") as output:
            while chunk := response.read(1024 * 1024):
                output.write(chunk)
    except (OSError, URLError) as exc:
        message = f"download VyOS ISO: {exc}"
        raise RuntimeError(message) from exc
    if target.stat().st_size == 0:
        message = f"download VyOS ISO: empty response for {url}"
        raise RuntimeError(message)
    return target


def _extract_squashfs_from_iso(iso_path: Path, work_dir: Path) -> Path:
    """Extract filesystem.squashfs from a VyOS ISO."""
    squashfs_path = work_dir / "filesystem.squashfs"
    if _which("7z"):
        _run(
            ["7z", "x", "-y", f"-o{work_dir}", str(iso_path), "live/filesystem.squashfs"],
            capture=True,
        )
        extracted = work_dir / "live" / "filesystem.squashfs"
        if extracted.exists():
            extracted.rename(squashfs_path)
            return squashfs_path
    if _which("xorriso"):
        _run(
            [
                "xorriso",
                "-osirrox",
                "on",
                "-indev",
                str(iso_path),
                "-extract",
                "/live/filesystem.squashfs",
                str(squashfs_path),
            ],
            capture=True,
        )
        if squashfs_path.exists():
            return squashfs_path
    message = "extract VyOS ISO: install 7z or xorriso"
    raise RuntimeError(message)


def _extract_squashfs_root(squashfs_path: Path, work_dir: Path) -> Path:
    """Extract squashfs into a rootfs directory."""
    if not _which("unsquashfs"):
        message = "extract VyOS rootfs: unsquashfs not found"
        raise RuntimeError(message)
    rootfs = work_dir / "rootfs"
    _run(["unsquashfs", "-f", "-d", str(rootfs), str(squashfs_path)], capture=True)
    return rootfs


def _import_vyos_rootfs(rootfs: Path) -> None:
    """Tar a rootfs and pipe it into podman import."""
    with (
        subprocess.Popen(  # nosec B603  # noqa: S603
            [_resolve_executable("tar"), "-C", str(rootfs), "-c", "."],
            stdout=subprocess.PIPE,
        ) as tar_process,
        subprocess.Popen(  # nosec B603  # noqa: S603
            [_resolve_executable("podman"), "import", "-", "vyos:latest"],
            stdin=tar_process.stdout,
        ) as import_process,
    ):
        if tar_process.stdout:
            tar_process.stdout.close()
        import_status = import_process.wait()
        tar_status = tar_process.wait()
    if tar_status != 0 or import_status != 0:
        message = f"import VyOS rootfs: tar={tar_status} podman={import_status}"
        raise RuntimeError(message)


def _prepare_vyos(arguments: argparse.Namespace) -> bool:
    """Prepare the vyos:latest image expected by the topology."""
    if arguments.skip_pull and _image_exists("vyos:latest"):
        return True
    if _image_exists("vyos:latest", dry_run=arguments.dry_run) and not arguments.dry_run:
        return True
    if not arguments.iso and _image_exists(arguments.image, dry_run=arguments.dry_run):
        _run(["podman", "tag", arguments.image, "vyos:latest"], dry_run=arguments.dry_run)
        return True
    if arguments.dry_run:
        _log.info("[dry-run] would prepare vyos:latest from %s", arguments.image)
        return True
    if not _which("unsquashfs") or not (_which("7z") or _which("xorriso")):
        _log.error("VyOS ISO extraction requires unsquashfs and either 7z or xorriso")
        return False
    with tempfile.TemporaryDirectory(prefix="vyos-bootstrap-") as temporary:
        work = Path(temporary)
        if arguments.iso:
            iso_path = Path(arguments.iso)
            if not iso_path.is_file():
                _log.error("VyOS ISO not found: %s", iso_path)
                return False
        else:
            version = arguments.version
            if version == "latest":
                version = _latest_vyos_version()
            iso_path = _download_vyos_iso(version, work)
        squashfs = _extract_squashfs_from_iso(iso_path, work)
        rootfs = _extract_squashfs_root(squashfs, work)
        _import_vyos_rootfs(rootfs)
    result = _run(
        [
            "podman",
            "run",
            "--rm",
            "vyos:latest",
            "/bin/bash",
            "-c",
            "cat /etc/vyos-release 2>/dev/null || echo 'VyOS image imported'",
        ],
        check=False,
        capture=True,
    )
    return result.returncode == 0


def _parse_loaded_image(stdout: str) -> str:
    """Parse an image reference from podman load output."""
    for line in stdout.splitlines():
        if "Loaded image" not in line:
            continue
        if "image(s):" in line:
            return line.split("image(s):", 1)[-1].strip()
        if "image:" in line:
            return line.split("image:", 1)[-1].strip()
    return ""


def _import_arista(arguments: argparse.Namespace) -> bool:
    """Import an operator-supplied Arista cEOS archive."""
    path = Path(arguments.archive)
    if not path.is_file():
        _log.error("Arista archive not found: %s", path)
        return False
    if _image_exists(arguments.tag, dry_run=arguments.dry_run):
        return True
    result = _run(
        ["podman", "load", "-i", str(path)],
        capture=True,
        check=False,
        dry_run=arguments.dry_run,
    )
    if result.returncode != 0:
        _log.error("podman load failed for Arista: %s", result.stderr.strip())
        return False
    loaded = _parse_loaded_image(result.stdout or "")
    if loaded and loaded != arguments.tag:
        tag_result = _run(
            ["podman", "tag", loaded, arguments.tag],
            check=False,
            dry_run=arguments.dry_run,
        )
        if tag_result.returncode != 0:
            return False
    return True


def _import_cisco_nested(path: Path) -> bool:
    """Extract and load a nested Docker archive from an XRd tarball."""
    with tempfile.TemporaryDirectory(prefix="xrd-bootstrap-") as temporary:
        work = Path(temporary)
        try:
            with tarfile.open(path) as archive:
                members = [
                    member
                    for member in archive.getmembers()
                    if member.name.endswith((".tar", ".tar.gz", ".tgz"))
                ]
                if not members:
                    return False
                archive.extract(members[0], work, filter="data")
                inner_path = work / members[0].name
        except (OSError, tarfile.TarError) as exc:
            _log.error("extract Cisco XRd archive: %s", exc)
            return False
        result = _run(
            ["podman", "load", "-i", str(inner_path)],
            capture=True,
            check=False,
        )
        return result.returncode == 0


def _import_cisco(arguments: argparse.Namespace) -> bool:
    """Import an operator-supplied Cisco XRd archive."""
    path = Path(arguments.archive)
    if not path.is_file():
        _log.error("Cisco XRd archive not found: %s", path)
        return False
    if _image_exists(arguments.tag, dry_run=arguments.dry_run):
        return True
    if arguments.dry_run:
        _log.info("[dry-run] would import Cisco XRd as %s", arguments.tag)
        return True
    result = _run(["podman", "load", "-i", str(path)], capture=True, check=False)
    if result.returncode == 0:
        return True
    return _import_cisco_nested(path)


def _parse_args() -> argparse.Namespace:
    """Parse the narrow vendor-image command line."""
    parser = argparse.ArgumentParser(description="Prepare containerlab vendor images.")
    parser.add_argument("-v", "--verbose", action="store_true")
    subcommands = parser.add_subparsers(dest="operation", required=True)

    vyos = subcommands.add_parser("vyos")
    vyos.add_argument("--version", default="latest")
    vyos.add_argument("--image", required=True)
    vyos.add_argument("--iso")
    vyos.add_argument("--skip-pull", action="store_true")
    vyos.add_argument("--dry-run", action="store_true")

    for name in ("arista", "cisco"):
        commercial = subcommands.add_parser(name)
        commercial.add_argument("--archive", required=True)
        commercial.add_argument("--tag", required=True)
        commercial.add_argument("--dry-run", action="store_true")
    return parser.parse_args()


def main() -> int:
    """Prepare exactly one requested vendor image."""
    arguments = _parse_args()
    logging.basicConfig(
        level=logging.DEBUG if arguments.verbose else logging.INFO,
        format="%(levelname)s %(message)s",
    )
    try:
        if arguments.operation == "vyos":
            success = _prepare_vyos(arguments)
        elif arguments.operation == "arista":
            success = _import_arista(arguments)
        else:
            success = _import_cisco(arguments)
    except (OSError, RuntimeError, subprocess.SubprocessError) as exc:
        _log.error("%s", exc)
        return 1
    return 0 if success else 1


if __name__ == "__main__":
    raise SystemExit(main())

File: test/internal/clabbootstrap/vendor_images.go
package clabbootstrap

import (
	"archive/tar"
	"compress/gzip"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"os"
	"path/filepath"
	"strings"
	"time"
)

const (
	vyosReleaseAPI    = "https://api.github.com/repos/vyos/vyos-rolling-nightly-builds/releases/latest"
	vyosDownloadBase  = "https://github.com/vyos/vyos-rolling-nightly-builds/releases/download"
	vyosTargetImage   = "vyos:latest"
	maxAPIResponse    = 1 << 20
	maxISOSize        = int64(16 << 30)
	maxNestedArchive  = int64(32 << 30)
	vendorHTTPTimeout = 10 * time.Minute
)

var (
	errVendorURL          = errors.New("vendor URL is not allowlisted")
	errVendorResponse     = errors.New("invalid vendor HTTP response")
	errVendorArchive      = errors.New("invalid vendor archive")
	errVendorImagePrepare = errors.New("vendor image preparation failed")
)

func prepareVyOS(ctx context.Context, options Options, runner Runner, sourceImage string) error {
	if options.SkipPull {
		exists, err := imageExists(ctx, options, runner, vyosTargetImage)
		if err != nil {
			return err
		}
		if exists {
			return nil
		}
	}
	exists, err := imageExists(ctx, options, runner, vyosTargetImage)
	if err != nil || exists {
		return err
	}
	if options.VyOSISO == "" {
		sourceExists, sourceErr := imageExists(ctx, options, runner, sourceImage)
		if sourceErr != nil {
			return sourceErr
		}
		if sourceExists || options.DryRun {
			return runCommand(ctx, runner, Command{
				Executable: executablePodman,
				Arguments:  []string{"tag", sourceImage, vyosTargetImage},
				DryRun:     options.DryRun,
			})
		}
	}
	if options.DryRun {
		return nil
	}
	return buildVyOSFromISO(ctx, options, runner)
}

func buildVyOSFromISO(ctx context.Context, options Options, runner Runner) error {
	work, err := os.MkdirTemp("", "gobfd-vyos-bootstrap-")
	if err != nil {
		return fmt.Errorf("create VyOS work directory: %w", err)
	}
	defer os.RemoveAll(work)

	isoPath, err := resolveVyOSISO(ctx, options, work)
	if err != nil {
		return err
	}

	squashfs, err := extractVyOSSquashFS(ctx, runner, isoPath, work)
	if err != nil {
		return err
	}
	rootfs := filepath.Join(work, "rootfs")
	if commandErr := runCommand(ctx, runner, Command{
		Executable: "unsquashfs",
		Arguments:  []string{"-f", "-d", rootfs, squashfs},
	}); commandErr != nil {
		return fmt.Errorf("extract VyOS rootfs: %w", commandErr)
	}
	containerfile := filepath.Join(work, "Containerfile.import")
	if writeErr := os.WriteFile(containerfile, []byte("FROM scratch\nADD rootfs/ /\n"), 0o600); writeErr != nil {
		return fmt.Errorf("write VyOS import Containerfile: %w", writeErr)
	}
	ignore := filepath.Join(work, ".containerignore")
	if writeErr := os.WriteFile(ignore, []byte("*\n!rootfs/**\n!Containerfile.import\n"), 0o600); writeErr != nil {
		return fmt.Errorf("write VyOS import ignore file: %w", writeErr)
	}
	if commandErr := runCommand(ctx, runner, Command{
		Executable: executablePodman,
		Arguments:  []string{"build", "--tag", vyosTargetImage, "--file", containerfile, work},
	}); commandErr != nil {
		return fmt.Errorf("build VyOS rootfs image: %w", commandErr)
	}
	result, err := runner.Run(ctx, Command{
		Executable: executablePodman,
		Arguments:  []string{"run", "--rm", vyosTargetImage, "cat", "/etc/vyos-release"},
	})
	if err != nil {
		return fmt.Errorf("verify VyOS image: %w", err)
	}
	if result.ExitCode != 0 {
		return fmt.Errorf("verify VyOS image: exit %d: %w", result.ExitCode, errVendorImagePrepare)
	}
	return nil
}

func resolveVyOSISO(ctx context.Context, options Options, work string) (string, error) {
	if options.VyOSISO != "" {
		if err := validateRegularFile(options.VyOSISO); err != nil {
			return "", fmt.Errorf("validate VyOS ISO %s: %w", options.VyOSISO, err)
		}
		return options.VyOSISO, nil
	}
	version := options.VyOSVersion
	if version == "latest" {
		var err error
		version, err = latestVyOSVersion(ctx)
		if err != nil {
			return "", err
		}
	}
	return downloadVyOSISO(ctx, version, work)
}

func extractVyOSSquashFS(ctx context.Context, runner Runner, isoPath, work string) (string, error) {
	target := filepath.Join(work, "filesystem.squashfs")
	sevenZip := Command{
		Executable: "7z",
		Arguments:  []string{"x", "-y", "-o" + work, isoPath, "live/filesystem.squashfs"},
	}
	if result, err := runner.Run(ctx, sevenZip); err == nil && result.ExitCode == 0 {
		extracted := filepath.Join(work, "live", "filesystem.squashfs")
		if err := os.Rename(extracted, target); err != nil {
			return "", fmt.Errorf("move VyOS squashfs: %w", err)
		}
		return target, nil
	}
	if err := runCommand(ctx, runner, Command{
		Executable: "xorriso",
		Arguments: []string{
			"-osirrox", "on", "-indev", isoPath,
			"-extract", "/live/filesystem.squashfs", target,
		},
	}); err != nil {
		return "", fmt.Errorf("extract VyOS ISO with xorriso: %w", err)
	}
	return target, nil
}

func importArista(ctx context.Context, options Options, runner Runner) error {
	if err := validateRegularFile(options.Archives.Arista); err != nil {
		return fmt.Errorf("validate Arista archive: %w", err)
	}
	exists, err := imageExists(ctx, options, runner, options.Tags.Arista)
	if err != nil || exists {
		return err
	}
	result, err := runner.Run(ctx, Command{
		Executable: executablePodman,
		Arguments:  []string{"load", "--input", options.Archives.Arista},
		DryRun:     options.DryRun,
	})
	if err != nil {
		return fmt.Errorf("load Arista archive: %w", err)
	}
	if result.ExitCode != 0 {
		return fmt.Errorf("load Arista archive: exit %d: %s: %w", result.ExitCode, result.Stderr, errVendorImagePrepare)
	}
	loaded := parseLoadedImage(result.Stdout)
	if loaded == "" || loaded == options.Tags.Arista {
		return nil
	}
	return runCommand(ctx, runner, Command{
		Executable: executablePodman,
		Arguments:  []string{"tag", loaded, options.Tags.Arista},
		DryRun:     options.DryRun,
	})
}

func importCisco(ctx context.Context, options Options, runner Runner) error {
	if err := validateRegularFile(options.Archives.Cisco); err != nil {
		return fmt.Errorf("validate Cisco archive: %w", err)
	}
	exists, err := imageExists(ctx, options, runner, options.Tags.Cisco)
	if err != nil || exists {
		return err
	}
	if options.DryRun {
		return nil
	}
	result, err := runner.Run(ctx, Command{
		Executable: executablePodman,
		Arguments:  []string{"load", "--input", options.Archives.Cisco},
	})
	if err != nil {
		return fmt.Errorf("load Cisco archive: %w", err)
	}
	if result.ExitCode == 0 {
		return nil
	}
	return importNestedCisco(ctx, runner, options.Archives.Cisco)
}

func importNestedCisco(ctx context.Context, runner Runner, archivePath string) error {
	work, err := os.MkdirTemp("", "gobfd-xrd-bootstrap-")
	if err != nil {
		return fmt.Errorf("create Cisco archive work directory: %w", err)
	}
	defer os.RemoveAll(work)
	inner, err := extractFirstNestedArchive(archivePath, work)
	if err != nil {
		return err
	}
	return runCommand(ctx, runner, Command{
		Executable: executablePodman,
		Arguments:  []string{"load", "--input", inner},
	})
}

func extractFirstNestedArchive(archivePath, destination string) (string, error) {
	file, err := os.Open(archivePath)
	if err != nil {
		return "", fmt.Errorf("open Cisco archive: %w", err)
	}
	defer file.Close()
	reader, closeReader, err := tarReader(file, archivePath)
	if err != nil {
		return "", err
	}
	target, extractErr := extractNestedArchive(reader, destination)
	return target, errors.Join(extractErr, closeReader())
}

func extractNestedArchive(reader *tar.Reader, destination string) (string, error) {
	for {
		header, nextErr := reader.Next()
		if errors.Is(nextErr, io.EOF) {
			return "", fmt.Errorf("cisco archive contains no nested image archive: %w", errVendorArchive)
		}
		if nextErr != nil {
			return "", fmt.Errorf("read Cisco archive: %w", nextErr)
		}
		name := strings.ToLower(header.Name)
		if header.Typeflag != tar.TypeReg ||
			!strings.HasSuffix(name, ".tar") && !strings.HasSuffix(name, ".tar.gz") && !strings.HasSuffix(name, ".tgz") {
			continue
		}
		if header.Size <= 0 || header.Size > maxNestedArchive {
			return "", fmt.Errorf("nested Cisco archive size %d: %w", header.Size, errVendorArchive)
		}
		target := filepath.Join(destination, "nested"+nestedArchiveSuffix(name))
		output, createErr := os.OpenFile(target, os.O_WRONLY|os.O_CREATE|os.O_EXCL, 0o600)
		if createErr != nil {
			return "", fmt.Errorf("create nested Cisco archive: %w", createErr)
		}
		_, copyErr := io.CopyN(output, reader, header.Size)
		closeErr := output.Close()
		if copyErr != nil {
			return "", fmt.Errorf("extract nested Cisco archive: %w", copyErr)
		}
		if closeErr != nil {
			return "", fmt.Errorf("close nested Cisco archive: %w", closeErr)
		}
		return target, nil
	}
}

func tarReader(file *os.File, name string) (*tar.Reader, func() error, error) {
	if strings.HasSuffix(strings.ToLower(name), ".gz") || strings.HasSuffix(strings.ToLower(name), ".tgz") {
		gzipReader, err := gzip.NewReader(file)
		if err != nil {
			return nil, func() error { return nil }, fmt.Errorf("open compressed Cisco archive: %w", err)
		}
		return tar.NewReader(gzipReader), gzipReader.Close, nil
	}
	return tar.NewReader(file), func() error { return nil }, nil
}

func nestedArchiveSuffix(name string) string {
	if strings.HasSuffix(name, ".tar.gz") {
		return ".tar.gz"
	}
	if strings.HasSuffix(name, ".tgz") {
		return ".tgz"
	}
	return ".tar"
}

func parseLoadedImage(output string) string {
	for line := range strings.Lines(output) {
		line = strings.TrimSpace(line)
		for _, marker := range []string{"Loaded image(s):", "Loaded image:"} {
			if _, value, found := strings.Cut(line, marker); found {
				return strings.TrimSpace(value)
			}
		}
	}
	return ""
}

func latestVyOSVersion(ctx context.Context) (string, error) {
	body, err := vendorGET(ctx, vyosReleaseAPI, maxAPIResponse)
	if err != nil {
		return "", fmt.Errorf("query latest VyOS release: %w", err)
	}
	var release struct {
		TagName string `json:"tag_name"`
	}
	if err := json.Unmarshal(body, &release); err != nil {
		return "", fmt.Errorf("decode latest VyOS release: %w", err)
	}
	if strings.TrimSpace(release.TagName) == "" {
		return "", fmt.Errorf("latest VyOS release lacks tag_name: %w", errVendorResponse)
	}
	return release.TagName, nil
}

func downloadVyOSISO(ctx context.Context, version, destination string) (string, error) {
	filename := "vyos-" + version + "-generic-amd64.iso"
	endpoint := vyosDownloadBase + "/" + url.PathEscape(version) + "/" + url.PathEscape(filename)
	target := filepath.Join(destination, filename)
	request, err := vendorRequest(ctx, endpoint)
	if err != nil {
		return "", err
	}
	client := vendorHTTPClient()
	response, err := client.Do(request)
	if err != nil {
		return "", fmt.Errorf("download VyOS ISO: %w", err)
	}
	defer response.Body.Close()
	if response.StatusCode != http.StatusOK {
		return "", fmt.Errorf("download VyOS ISO: status %s: %w", response.Status, errVendorResponse)
	}
	if response.ContentLength > maxISOSize {
		return "", fmt.Errorf("download VyOS ISO: content length %d: %w", response.ContentLength, errVendorResponse)
	}
	file, err := os.OpenFile(target, os.O_WRONLY|os.O_CREATE|os.O_EXCL, 0o600)
	if err != nil {
		return "", fmt.Errorf("create VyOS ISO: %w", err)
	}
	written, copyErr := io.Copy(file, io.LimitReader(response.Body, maxISOSize+1))
	closeErr := file.Close()
	if copyErr != nil {
		return "", fmt.Errorf("write VyOS ISO: %w", copyErr)
	}
	if closeErr != nil {
		return "", fmt.Errorf("close VyOS ISO: %w", closeErr)
	}
	if written <= 0 || written > maxISOSize {
		return "", fmt.Errorf("download VyOS ISO: size %d: %w", written, errVendorResponse)
	}
	return target, nil
}

func vendorGET(ctx context.Context, endpoint string, limit int64) ([]byte, error) {
	request, err := vendorRequest(ctx, endpoint)
	if err != nil {
		return nil, err
	}
	response, err := vendorHTTPClient().Do(request)
	if err != nil {
		return nil, fmt.Errorf("send GET %s: %w", endpoint, err)
	}
	defer response.Body.Close()
	if response.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("GET %s: status %s: %w", endpoint, response.Status, errVendorResponse)
	}
	body, err := io.ReadAll(io.LimitReader(response.Body, limit+1))
	if err != nil {
		return nil, fmt.Errorf("read %s: %w", endpoint, err)
	}
	if int64(len(body)) > limit {
		return nil, fmt.Errorf("GET %s exceeds %d bytes: %w", endpoint, limit, errVendorResponse)
	}
	return body, nil
}

func vendorRequest(ctx context.Context, endpoint string) (*http.Request, error) {
	if err := validateVendorURL(endpoint); err != nil {
		return nil, err
	}
	request, err := http.NewRequestWithContext(ctx, http.MethodGet, endpoint, nil)
	if err != nil {
		return nil, fmt.Errorf("create vendor request: %w", err)
	}
	request.Header.Set("Accept", "application/vnd.github+json")
	request.Header.Set("User-Agent", "gobfd-bootstrap")
	return request, nil
}

func vendorHTTPClient() *http.Client {
	return &http.Client{
		Timeout: vendorHTTPTimeout,
		CheckRedirect: func(request *http.Request, _ []*http.Request) error {
			return validateVendorURL(request.URL.String())
		},
	}
}

func validateVendorURL(endpoint string) error {
	parsed, err := url.Parse(endpoint)
	if err != nil {
		return fmt.Errorf("parse vendor URL: %w", err)
	}
	allowedHost := parsed.Hostname() == "api.github.com" || parsed.Hostname() == "github.com" ||
		parsed.Hostname() == "release-assets.githubusercontent.com"
	if parsed.Scheme != "https" || !allowedHost || parsed.User != nil ||
		(parsed.Port() != "" && parsed.Port() != "443") {
		return fmt.Errorf("%w: %q", errVendorURL, endpoint)
	}
	return nil
}

func validateRegularFile(path string) error {
	info, err := os.Stat(path)
	if err != nil {
		return fmt.Errorf("stat %s: %w", path, err)
	}
	if !info.Mode().IsRegular() {
		return fmt.Errorf("path %s is not a regular file: %w", path, errVendorArchive)
	}
	return nil
}
